liteeth/core/ip.py

```diff
@@ -145,7 +158,21 @@ def __init__(self, mac_address, ip_address, arp_table, dw=8):
                 )
             )
         )
-        self.comb += arp_table.request.ip_address.eq(sink.ip_address)
+
+        if gateway is not None:
+            masked_ip = Signal(32, reset_less=True)
+            in_subnet = Signal(1, reset_less=True)
+
+            self.sync += masked_ip.eq(sink.ip_address & netmask)
+            self.sync += in_subnet.eq(masked_ip == netaddress)
+            self.sync += If(in_subnet,
```

Comment:
@enjoy-digital  Now having slept on it, I'm not actually sure this will correctly. Essentially the IP address now needs to be stable for 3 clock cycles before the `SEND_MAC_ADDRESS_REQUEST` is reached. I don't know enough about packetizers and entire stack whether this is true. We have a single cycle for sure so I could shrink the pipeline.